src/middlewared/middlewared/plugins/vm.py

```diff
@@ -248,8 +245,15 @@ def __init__(self, manager, vm):
 
         while True:
             line = await self.proc.stdout.readline()
+
+            if self.proc.returncode:
+                await self.__teardown_guest_vmemory(self.vm['id'])
+                await self.destroy_vm()
+                raise CallError(line.decode())
```

Comment:
This seems wrong, this is already handle below, we need to figure out what is wrong there.

Reply:
This isn't handled below. This is for issuse *before* bhyve, like no backing zvol. Bhyve will throw an error code of 1 for this (just shuts the vm down) and no error will be raised, as it's a valid code for just shutting the vm down. See line 275. 

Here's a snapshot of the log for you:
```
[2018/07/02 21:39:32] (DEBUG) VMService.run():154 - ====> NIC_ATTACH: vtnet0
[2018/07/02 21:39:32] (DEBUG) VMService.run():231 - Starting bhyve: bhyve -A -H -w -c 1 -m 512 -s 0:0,hostbridge -s 31,lpc -l com1,/dev/nmdm14A -l bootrom,/usr/local/share/uefi-firmware/BHYVE_UEFI.fd -s 3,e1000,tap26,mac=00:a0:98:6c:53:0e -s 4,ahci-hd,/dev/zvol/tank/testFOO-7f7f0m -s 5,ahci-cd,/mnt/tank/openSUSE-Leap-15.0-DVD-x86_64-Current.iso -s 29,fbuf,vncserver,tcp=192.168.122.111:5643,w=1024,h=768,,wait -s 30,xhci,tablet 14_testFOO
[2018/07/02 21:39:32] (TRACE) middlewared.send_event():1054 - Sending event "ADDED":{'data': {'system': 'DEVFS', 'subsystem': 'CDEV', 'type': 'CREATE', 'cdev': 'tap26'}}
[2018/07/02 21:39:32] (TRACE) middlewared.send_event():1054 - Sending event "ADDED":{'data': {'system': 'IFNET', 'subsystem': 'tap26', 'type': 'ATTACH'}}
[2018/07/02 21:39:32] (DEBUG) VMService.run():249 - ==> Start WEBVNC at port 5543 with pid number 43057
[2018/07/02 21:39:32] (DEBUG) VMService.run():262 - testFOO: bhyve: Could not open backing file: /dev/zvol/tank/testFOO-7f7f0m: No such file or directory

[2018/07/02 21:39:32] (INFO) VMService.run():277 - ===> Powered off VM: testFOO ID: 14 BHYVE_CODE: 1
[2018/07/02 21:39:32] (TRACE) middlewared.send_event():1054 - Sending event "ADDED":{'data': {'system': 'ETHERNET', 'subsystem': 'tap26', 'type': 'IFATTACH'}}
[2018/07/02 21:39:32] (TRACE) middlewared.send_event():1054 - Sending event "ADDED":{'data': {'system': 'DEVFS', 'subsystem': 'CDEV', 'type': 'CREATE', 'cdev': 'vmm/14_testFOO'}}
[2018/07/02 21:39:32] (TRACE) middlewared.send_event():1054 - Sending event "ADDED":{'data': {'system': 'IFNET', 'subsystem': 'tap26', 'type': 'LINK_UP'}}
[2018/07/02 21:39:32] (ERROR) VMService.running():409 - ===> VMM testFOO is running without bhyve process.
[2018/07/02 21:39:32] (WARNING) VMService.destroy_vm():290 - ===> Destroying VM: testFOO ID: 14 BHYVE_CODE: 1
[2018/07/02 21:39:32] (DEBUG) VMService.kill_bhyve_web():374 - ==> Killing WEBVNC: 43057
```

You'll notice the real error never raises, we just destroy the same resources and return True as in this case it's meant to under normal circumstances with this exit code.

Reply:
That does make much sense to me.

You are literally checking for the returncode on every iteration and tearing down memory and destroying vm, I dont see how exactly the same could not be done below.

Quite the opposite, doing it at that point will screw up all the logic below.

If bhyve exit code cannot differentiate a clean poweroff and and other errors like no backing zvol we need to either workaround it by grepping stderr or ask @araujobsd for another way to detect this (a different exit code or some such).

Reply:
bhyve can return the following exit status to indicates how the VM was terminated:
     0       rebooted
     1       powered off
     2       halted
     3       triple fault

It cannot detect other errors such like no disks and etc.

Reply:
@araujobsd Is it impossible to add another error code for user configuration error? (e.g. missing zvol, nic card, etc)

Reply:
It is not ideal to add another error check inside bhyve without a generic interface to deal with all bhyve devices. Usually the projects around bhyve such like iohyve, vm-bhyve and etc handle the error prior to launch bhyve.

Reply:
It is under review https://reviews.freebsd.org/D16161

Reply:
Awesome, thank you!

Reply:
Thanks!

Reply:
@araujobsd Thank you. Any chance it can be merged to freenas/os earlier? I have been told we will be in a freeze again next week and it would be great to have this resolved before the freeze, we have several tickets about this same problem.

Reply:
I can push it tomorrow then!

Reply:
Done: https://github.com/freenas/os/commit/c8ca6b483b2ceabce79a51e6bd2bb775172f78a7